File: 17. Letter Combinations of a Phone Number/Rewriting Array/letter_combinations_of_a_phone_number.rb
$combinations = {
	0=> " ",
	1=> "",
	2=> "abc",
	3=> "def",
	4=> "ghi",
	5=> "jkl",
	6=> "mno",
	7=> "pqrs",
	8=> "tuv",
	9=> "wxyz"
	}
	
def letter_combinations (digits)
	out = []
	digits = String(digits)
	
	for dig in 0...digits.length
	#digits.each do |d|
		d = dig
		key_letters = $combinations[Integer(digits[d])]
		
		if ((out == []) and (out.length == 0))
			for i in 0...key_letters.length
			#key_letters.each do |letter|
				letter = key_letters[i]
			
				out << letter
			end
		else
			holder = []
			for i in 0...key_letters.length
			#key_letters.each do |letter|
				letter = key_letters[i]
				#for out_words in out
				out.each do |out_words|
					holder << (out_words + letter)
				end
			end
			
			out = holder
		end
	end
	
	out.sort()
	
	return out
end

digits = "234"

puts (letter_combinations(digits));

File: 17. Letter Combinations of a Phone Number/Rewriting Array/letter_combinations_of_a_phone_number.js
let combinations = {
	0: " ",
	1: "",
	2: "abc",
	3: "def",
	4: "ghi",
	5: "jkl",
	6: "mno",
	7: "pqrs",
	8: "tuv",
	9: "wxyz"
	}
	
function letterCombinations (digits){
	let out = [];
	digits = String(digits);
	//console.log(digit);
	for (dig of digits){
		//console.log(dig);
		let keyLetters = combinations[Number(dig)];
		//console.log(keyLetters);
		if ((out.length == 0)){ //(out == []) && 
			for (letter of keyLetters){
				out.push(letter);
			}
		} else {
			let holder = []
			for (letter of keyLetters){
				for (outWords of out){
					holder.push(outWords + letter);
				}
			}
			
			out = holder;
		}
	}
	
	out.sort();
	
	return (out);
}

let digits = "239";
//console.log (letterCombinations(digits));

let c = letterCombinations(digits);

for (a of c){
	console.log(a);
}
